src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -1667,6 +1693,69 @@ protected override void HandleModulesRequestAsync(IRequestResponder<ModulesArgum
             responder.SetResponse(response);
         }
 
+        // test -- need to delete
+        protected override void HandleDisassembleRequestAsync(IRequestResponder<DisassembleArguments, DisassembleResponse> responder)
+        {
+            DisassembleResponse response = new DisassembleResponse();
+            DisassembleArguments disassembleArguments = responder.Arguments;
+            Debug.Assert(!string.IsNullOrEmpty(disassembleArguments.MemoryReference));
+            try
+            {
+                if (GetMemoryContext(disassembleArguments.MemoryReference, disassembleArguments.Offset, out IDebugMemoryContext2 memoryContext, out ulong address) == HRConstants.S_OK)
+                {
+                    IDebugCodeContext2 codeContext = memoryContext as IDebugCodeContext2;
+                    if (!(codeContext is IDebugCodeContext2))
+                    {
+                        throw new AD7Exception(AD7Resources.Error_ConvertMemoryContext);
+                    }
+                    if (m_program.GetDisassemblyStream(enum_DISASSEMBLY_STREAM_SCOPE.DSS_ALL, codeContext, out IDebugDisassemblyStream2 disassemblyStream) == HRConstants.S_OK)
+                    {
+                        if (disassembleArguments.InstructionOffset != 0)
+                        {
+                            if (disassemblyStream.Seek(enum_SEEK_START.SEEK_START_BEGIN, codeContext, address, (long)disassembleArguments.InstructionOffset) != HRConstants.S_OK)
+                            {
+                                throw new AD7Exception(AD7Resources.Error_SeekDisassemblyStream);
+                            }
+                        }
+                        DisassemblyData[] prgDisassembly = new DisassemblyData[disassembleArguments.InstructionCount];
+                        if (disassemblyStream.Read((uint)disassembleArguments.InstructionCount, enum_DISASSEMBLY_STREAM_FIELDS.DSF_ALL, out uint pdwInstructionsRead, prgDisassembly) == HRConstants.S_OK)
+                        {
+                            Debug.Assert(disassembleArguments.InstructionCount == pdwInstructionsRead);
+                            foreach (DisassemblyData data in prgDisassembly)
+                            {
+                                if (data.dwFlags.HasFlag(enum_DISASSEMBLY_FLAGS.DF_HASSOURCE))
+                                {
+                                    Debug.Fail("Warning: engine supports mixed instruction/source disassembly, but OpenDebugAD7 does not.");
```

Comment:
Would this be better off as a logged warning instead of a Debug.Fail? The end user won't see this message and you as the developer of MIEngine/OpenDebugAD7 will only see this with debug builds.

Reply:
That seems fine to me -- this is here so that if we add support in the MIEngine (or any other engine we might use in the future), and are wondering why it doesn't work in OpenDebugAD7, we will know right away.